packages/swingset-liveslots/src/virtualObjectManager.js

```diff
@@ -15,137 +15,275 @@ const { details: X, quote: q } = assert;
+    return 'not';
+  }
+  let established;
+  for (const prop of Reflect.ownKeys(obj)) {
+    const value = obj[prop];
+    let current;
+    if (typeof value === 'function') {
+      current = 'one';
+    } else if (
+      !inner &&
+      typeof value === 'object' &&
+      assessFacetiousness(value, true) === 'one'
+    ) {
+      if (typeof prop === 'symbol') {
+        // can't have symbol-named facets
+        return 'not';
       }
-      return innerObj;
-    },
+      current = 'many';
+    } else {
+      return 'not';
+    }
+    if (!established) {
+      established = current;
+    } else if (established !== current) {
+      return 'not';
+    }
+  }
+  if (!established) {
+    // empty objects are methodless Far objects
+    return 'one';
+  } else {
+    return /** @type {Facetiousness} */ (established);
+  }
+}
+
+// To manage context and state and data correctly (not sensitive to GC), we need
+// two Caches. The first is "dataCache", and maps baseRef to state data: {
+// capdata, valueMap }.  "capdata" is a record like { propX: capdataX, propY:
+// capdataY }, one property per property name of the "state" object, initialized
+// from 'initialData' when the instance was first created. "valueMap" is a Map,
+// from "propX" to the unserialized value of `state.propX`. The second cache is
+// "contextCache", and maps baseRef to a context object, which is either {
+// state, self } or { state, facets } depending on the facetiousness of the
+// VO. Here, "state" is an object with one accessor pair (getter+setter) per
+// state property name.
+//
+// The "state" getters/setters know which baseRef they should use. When invoked,
+// they pull the state data from dataCache.get(baseRef).valueMap. The setter
+// will modify valueMap in place and mark the entry as dirty, so it can be
+// serialized and written back at end-of-crank.
+//
+// Each Representative is built as an Exo with defendPrototype (cohorts of
+// facets are built with defendPrototypeKit). These are given a "contextMap" for
+// later use. For each facet, they build a prototype object with wrappers for
+// all the methods of that particular facet. When those wrappers are invoked,
+// the first thing they do is to look up "this" (the facet/Representative) in
+// "contextMap" to obtain the "context" object: { state, self } or { state,
+// facets }.
+//
+// Our GC sensitivity contraints are:
+// * userspace must not be able to sense garbage collection
+// * Representatives are created on-demand when userspace deserializes a vref
+//   * they disappear when UNREACHABLE and GC collects them
+// * syscalls must be a deterministic function of userspace behavior
+//   * that includes method invocation and "state" property read/writes
+//   * they should not be influenced by GC until a bringOutYourDead delivery
+//
+// Userspace might hold on to a Representative, the context object, or the state
+// object, for an unbounded length of time: beyond a single crank/delivery. They
```

Comment:
There's a larger comment about that down by `makeRepresentative` (I think I wrote up the constraints twice). I'll remove the second paragraph here and refer to the later comments for the details.

The short answer is that `self` is the Representative: userspace can hold it strongly, but the lose the RAM-saving benefits if they do (as we've learned in the last few days, this is surprisingly easy to do). Storing them in the WeakMap/WeakSet that userspace gets (i.e. `VirtualObjectAwareWeakMap/Set`) holds the Representative strongly: those two collections are not virtualized. Storing them in a properly virtual collection (e.g. BigMapStore) will remember the vref but allow the Representative to drop out of RAM.

`facets` is the unusual one, because it shares a lifetime with the actual facets. It's marked as "unweakable", which means that if you store it in a VOAwareWeakSet, it will actually be held strongly, denying its use as a sensor.

Reply:
> Storing them in the WeakMap/WeakSet that userspace gets (i.e. `VirtualObjectAwareWeakMap/Set`) holds the Representative strongly: those two collections are not virtualized.

To clarify, storing them as value in a WeakMap holds them strongly. If used as keys in a `VirtualObjectAwareWeakMap`, or as keys/values in a `VirtualObjectAwareWeakSet`, then they can be paged out.

Reply:
> `facets` is the unusual one, because it shares a lifetime with the actual facets. It's marked as "unweakable", which means that if you store it in a VOAwareWeakSet, it will actually be held strongly, denying its use as a sensor.

I'm still wondering if it wouldn't be cleaner to get rid of the `unweakable` stuff, and simply represent `facets` as the bare base ref.